compiler+runtime/src/cpp/jank/codegen/llvm_processor.cpp

```diff
@@ -954,6 +953,76 @@ namespace jank::codegen
     return call;
   }
 
+  llvm::Value *llvm_processor::gen(expr::case_<expression> const &expr,
+                                   expr::function_arity<expression> const &arity)
+  {
+    auto const current_fn(ctx->builder->GetInsertBlock()->getParent());
+    auto const position{ expr.position };
+    auto const value(gen(expr.value_expr, arity));
+    auto const is_return{ position == expression_position::tail };
+    auto const integer_fn_type(llvm::FunctionType::get(
+      ctx->builder->getInt64Ty(),
+      { ctx->builder->getPtrTy(), ctx->builder->getInt64Ty(), ctx->builder->getInt64Ty() },
+      false));
+    auto const fn(ctx->module->getOrInsertFunction("shift_mask_case_integer", integer_fn_type));
+    llvm::SmallVector<llvm::Value *, 3> const args{
+      value,
+      llvm::ConstantInt::getSigned(ctx->builder->getInt64Ty(), expr.shift),
+      llvm::ConstantInt::getSigned(ctx->builder->getInt64Ty(), expr.mask)
+    };
+    auto const call(ctx->builder->CreateCall(fn, args));
+    auto const switch_val(ctx->builder->CreateIntCast(call, ctx->builder->getInt64Ty(), true));
+    auto const default_block = llvm::BasicBlock::Create(*ctx->llvm_ctx, "default", current_fn);
+    auto const switch_
+      = ctx->builder->CreateSwitch(switch_val, default_block, expr.transformed_keys.size());
+    auto const merge_block
+      = is_return ? nullptr : llvm::BasicBlock::Create(*ctx->llvm_ctx, "merge", current_fn);
+
+    ctx->builder->SetInsertPoint(default_block);
+    auto const default_val{ gen(expr.default_expr, arity) };
+    if(!is_return)
+    {
+      ctx->builder->CreateBr(merge_block);
+    }
+    auto const default_block_exit = ctx->builder->GetInsertBlock();
+
+    llvm::SmallVector<llvm::BasicBlock *> case_blocks;
+    llvm::SmallVector<llvm::Value *> case_values;
+    for(size_t block_counter = 0; block_counter < expr.transformed_keys.size(); ++block_counter)
+    {
+      auto const block_name = fmt::format("case_{}", block_counter);
+      auto const block{ llvm::BasicBlock::Create(*ctx->llvm_ctx, block_name, current_fn) };
+      switch_->addCase(llvm::ConstantInt::getSigned(ctx->builder->getInt64Ty(),
+                                                    expr.transformed_keys[block_counter]),
+                       block);
+
+      ctx->builder->SetInsertPoint(block);
+      auto const case_val{ gen(expr.exprs[block_counter], arity) };
+      case_values.push_back(case_val);
+      if(!is_return)
+      {
+        ctx->builder->CreateBr(merge_block);
+      }
+      case_blocks.push_back(ctx->builder->GetInsertBlock());
+    }
+
+    // Handle value position with PHI node
```

Comment:
Please use `/* */` comments and end them with a period.

Also, for comments, please explain the WHY more so than the WHAT. So, we're handling value position, but why does that need special handling?

Reply:
removed. that's my note to myself that i forgot to remove.